Post Editor: Redirect to Gutenberg routes if the site is opted in.

kwight opened the pull request:
**DO NOT MERGE – may be replaced with #28216**

We are adding an endpoint for checking if a site has been opted in to Gutenberg. This knowledge will be needed for the standard edit routes: if the site is opted in, we will need to redirect to the Gutenberg-ized route, and before displaying any editor UI – else, continue loading the "classic" editor.

Fixes #27431 .

kwight committed:
Gutenberg: Redirect to Gutenberg routes if the site is opted in.

kwight commented on client/post-editor/controller.js:
```diff
@@ -253,4 +254,12 @@ export default {
+	gutenberg: function( context, next ) {
+		const siteId = getSelectedSiteId( context.store.getState() );
+		context.store.dispatch( listBlogStickers( siteId ) );
+		// wait until we have the site's blog stickers;
```
This is where I'm at a loss how to proceed :)

jsnajdr commented:
I assume you want to `page.redirect` to Gutenberg route if the blog sticker says so. Then the page.js handler is a good place to do it.

You need a handler that waits until the blog stickers for the site are available, triggers a fetch if needed, and then either redirects, or passes control to the next handler in the handler chain.

```js
async function redirectToGutenbergIfStickerSaysSo( context, next ) {
  const siteId = getSelectedSiteId( context.store.getState() );
  const blogStickers = await queryBlogStickers( context.store, siteId );
  if ( blogStickers.includes( 'enable-gutenberg' ) ) {
    page.redirect( toGutenbergRoute( context.path ) );
    return;
  }
  
  next();
}
```

Now the `queryBlogStickers` function should fetch the stickers and fulfill a promise when they are available:
```js
function queryBlogStickers( store, siteId ) {
  return new Promise( resolve => {
    // if stickers are already in Redux, fulfill immediately
    const stickers = getBlogStickers( store.getState(), siteId );
    if ( stickers ) {
      resolve( stickers );
      return;
    }

    // issue a fetch request
    store.dispatch( listBlogStickers( siteId ) );

    // subscribe to store and wait until stickers arrive
    const unsubscribe = store.subscribe( () => {
      const stickers = getBlogStickers( store.getState(), siteId );
      if ( stickers ) {
        unsubscribe();
        resolve( stickers );
      }
    } );
  } );
}
```

kwight commented:
@jsnajdr Cool, yes, thank you! @dmsnell and I chatted more today, and should have a way forward in place early next week 👍

kwight committed:
Update some imaginable updates once the "opt-in" API endpoint is in
place and active.

kwight commented:
@dmsnell Another imagining of what could be, after our discussions today.

dmsnell committed:
httpData: Promise Promise interface to wait on data fulfillment

The data layer and `httpData` are systems built upon synchronous
data flows coupled by introspectable Redux _descriptions of requests_.
While this model provides valuable benefits to a React+Redux environment
there are negative tradeoffs that come about when people want to use
it in imperative places that _must_ wait on data fulfillment before
proceeding with app execution.

This new helper function `withData()` is an addition to the `httpData`
system that adds a `Promise` interface for data requests to allow this
sequential and blocking operation.

This should be used when performing server-side rendering and when
a `controller` depends on remote data to make decisions about routing
and paths. It should _not_ be used inside React components.

blowery commented on client/post-editor/controller.js:
```diff
@@ -253,4 +253,14 @@ export default {
+
+	gutenberg: async function( context, next ) {
+		const siteId = await getSelectedSiteId( context.store.getState() );
+		const editor = await getSelectedEditor( siteId );
```
do you need a try / finally to wrap these? If they reject and throw, you'll never call next and end up showing nothing.

blowery commented on client/post-editor/controller.js:
```diff
@@ -253,4 +253,14 @@ export default {
 	},
+
+	gutenberg: async function( context, next ) {
+		const siteId = await getSelectedSiteId( context.store.getState() );
```
this isn't an async function or something that returns a promise. Therefore, no need for the `await`.

blowery commented:
Is this tied to #27848 at all?

kwight commented:
Sorry for the confusion everyone – this is now just a work-in-progress PR of an approach @dmsnell and I have talked about, definitely not ready for merge proposing 🙃

dmsnell committed:
adjust from PR feedback

kwight committed:
First pass at using `waitForData`.

kwight committed:
Tweak waitForData() call. `requestSelectedEditor` definition to come.

kwight committed:
Add `requestSelectedEditor`.

kwight commented on client/state/data-getters/index.js:
```diff
@@ -176,6 +176,20 @@ export const requestGutenbergDemoContent = () =>
+				method: 'GET',
+				apiNamespace: 'wpcom/v2',
+			},
+			{}
```
This is as far as I got @dmsnell ; I'm thinking I'm missing `setSelectedEditor` here (although I don't see how to pass its params in the way it would need)?

dmsnell replied:
```js
( siteId, editor ) => http( {
  path: …,
  method: …,
  onSuccess: setEditor( siteId, editor )
} )
```

kwight committed:
Simplify `next()` handling.

kwight committed:
Make the request ID unique.

kwight committed:
Add an `onSuccess` handler that uses the non-existant
`setSelectedEditor` (coming from #27892).

kwight commented on client/post-editor/controller.js:
```diff
@@ -81,7 +83,7 @@ function getPressThisContent( query ) {
 				<p>
 					<a href={ url }>
-						<img src={ image } />
+						<img src={ image } alt="" />
```
It was a simple solution to a required lint fix; it's valid, although I could give it a slightly more helpful "featured image for the post" or something like that.

dmsnell replied:
😉 this is an example of a "fix" working around the problem instead of fixing it.

I recommend either _being bold_ and telling `eslint` "NO!" or fixing it by providing a more helpful alternate value. I'm not familiar with the code so I don't know what would be appropriate, but I'm pretty sure an empty string is not better than a missing attribute

jsnajdr replied:
`alt=""` has a well-defined meaning that is different from not specifying the attribute at all: it tells the screenreader that the image is for decoration and can be safely ignored by it.

See also:
https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/alt-text.md
https://www.w3.org/WAI/WCAG21/Techniques/html/H67

jsnajdr replied:
`alt=""` has a well-defined meaning that is different from not specifying the attribute at all: it tells the screenreader that the image is for decoration and can be safely ignored by it.

See also:
https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/alt-text.md
https://www.w3.org/WAI/WCAG21/Techniques/html/H67

jsnajdr commented on client/post-editor/index.js:
```diff
@@ -15,18 +15,39 @@ import { makeLayout, render as clientRender } from 'controller';
+		'/post/:site?/:post?',
+		siteSelection,
+		controller.post,
+		controller.gutenberg,
```
`controller.gutenberg` should run before `controller.post` -- do the redirect before starting to show a classic editor for the post, not after.

kwight committed:
Correct order of callbacks.

kwight commented:
Replacing with #28216 .
